docs/kittengrad.md

```diff
@@ -0,0 +1,151 @@
+a p2p gossip protocol.
+
+### Transactor State
+
+Each transactor maintains a local copy of the blockchain on disk, and a pointer
+to the last commited block. The transactor also maintains on disk the mediachain
+index which corresponds to the blockchain, mapping canonical to chain references.
+The index is used for creating update transactions and providing read access
+to the mediachain.
+
+In addition to transactions commited in the blockchain, each transactor
+maintains a worklog of pending transactions. The blockchain index together with
+the worklog transactions compute an aggregate index, the working index.
+
+Read requests are serviced by the working index, while write operations are
+initiated on the working index and completed through the transaction protocol.
+
+### Transaction Protocol
+#### Transaction Initiation
+
+In order to perform a write operation, the transactor checks the validity
+of the operation against the working index. Duplicate inserts are rejected,
+as are updates for unknown canonicals. It also checks its worklog backlog,
+against some safety threshold.
+
+The transactor then generates a transaction id and adds the data to the
+datastore and the transaction to the worklog. It then brodcasts the
+transaction to the network, and returns the transaction id to the client.
+
+If the transaction is the first one in the worklog, the transactor sets
+a timer that ensures a block confirming the transaction will eventually
+be commited with the Block Commit Protocol.
+If the worklog size exceeds the maximum block size, and there is no
+commit in progress, the transactor initiates the Block Commit Protocol
+in order to commit some transactions off its worklog.
+
+#### Transaction Propagation
+
+Each transactor receiving a transaction, validates it against its
+working index and, if valid, adds it to the worklog. This ensures
+that each transactor has a valid working index from which to generate
+a block. Conflicting writes will result in invalid transactions, which
+will be rejected and not added to any block proposed by the rejecting
+transactor during block commit.
+
+#### Transaction Commit
+
+When the Block Commit Protocol completes, it results into a block
+ready for commit. The block is scanned for confirming pending transactions,
+which are removed from the worklog.
+
+The block is stored and becomes the current blockchain pointer, while
+the blockchain index is updated in place. The worklog of remaining pending
+transactions is then revalidated against the blockchain index.
+Invalid transactions are rejected, and it is the responsibility of the originating
+transactor to replay them.
+
+In this way, the block commit provides a conflict resolution mechanism
+for concurrent updates. Inserts may use a deterministic transaction id,
+based on the canonical reference, which allows merging and confirmation
+of duplicate inserts without a replay. Updates however are dependent
```

Comment:
I originally had a section for a potential CRDT solution using sets, but they do have to be eventually written out during block proposal. The difference is that the entire data of the transaction has to be passed around through the transactor network instead of a datastore reference.